src/Core/Client/Client.php

```diff
@@ -279,11 +280,31 @@ class Client extends Configurable implements ClientInterface
      *
      * If an EventDispatcher instance is provided this will be used instead of creating a new instance
      *
-     * @param array                    $options
-     * @param EventDispatcherInterface $eventDispatcher
+     * @param AdapterInterface              $adapter
+     * @param array|null                    $options
+     * @param EventDispatcherInterface|null $eventDispatcher
      */
-    public function __construct(array $options = null, EventDispatcherInterface $eventDispatcher = null)
+    public function __construct($adapter = null, $options = null, $eventDispatcher = null)
```

Comment:
Should we allow passing null then for `$adapter` and fall back to instantiate a `Curl` adapter?

Should the constructor signature then be

```php
public function __construct(array $options = null, EventDispatcherInterface $eventDispatcher = null, AdapterInterface $adapter = null)
```

Reply:
If the constructor signature changes, a new optional argument is best added as last. Making it the first argument in the new signature would break backward compatibility.

Reply:
I think we should keep the "dualism" @dmircher introduced in the code below. The first parameter could be an adapter instance or an options array. We already use such a pattern in the code base. But we should name the parameter  something like $adapter_or_options and the type needs to be AdapterInterface|array|null

On null we should keep the existing behaviour and use the curl adapter.

Reply:
this can turn ugly very quickly, but i'm with @mkalkbrenner and @dmaicher to have the ``$adapter`` parameter as the first argument...

not entirely sure whether it would be added value to rename the parameter to something different and in the end change it back to ``$adapter`` as this is supposed to be a temporary solution. i'd opt for a user deprecation error in the bit below where ``if (is_array($adapter)) {`` returns true.

Reply:
So I kept the adapter as the first argument. In Solarium 6 it should be this signature then:

```php
public function __construct(AdapterInterface $adapter = null, array $options = null, EventDispatcherInterface $eventDispatcher = null)

Reply:
i'm a bit on the fence about this one...

i don't think the ``AdapterInterface`` should be nullable as the client can't function without it and i'm not entirely sure falling back to a default should be the task of the client itself.

same story for the ``EventDispatcherInterface``; all essential methods (``createRequest``, ``createResult``, ``execute`` & ``executeRequest``) in the client expect the event dispatcher not to be null. to add insult to injury, solarium listens to itself: the ``PostBigRequest`` won't function with the absence of an event dispatcher. once more i don't think defining the dispatcher should be tasked to the client.

which leaves us with the ``$options`` parameter being the only optional one.
so personally i'd prefer the following signature:

```php
public function __construct(AdapterInterface $adapter, EventDispatcherInterface $eventDispatcher, array $options = [])
```

this would require some additional deprecation warnings and of course has a bit more of an impact, but if we're changing constructor signatures i think a discussion about it prevents us from doing it more than once.

so, any thoughts?

Reply:
@wickedOne I see your point. But I think my suggestion in https://github.com/solariumphp/solarium/pull/754#discussion_r399118932 is a good intermediate step for 5.2 as it is fully compatible for existing installations and allows us to fire deprecation warnings.
For 6.0 we can continue to discuss the API change as 5.2 opens the window to announce it.

Reply:
Ok I agree here. So now the constructor should be fully backwards compatible with deprecations.

In Solarium 6 we can change it then with proper type hints.

Note: I did not change the nullable event dispatcher argument. I think this should be done in the scope of some other PR before we tag a new release for 5.2 then.

Reply:
@mkalkbrenner yes i agree with your comment and the solution provided by @dmaicher 
my previous comment in this thread was about the final signature for solarium 6.x

changes in this pr should enable us to add appropriate deprecation warnings for the final signare as we're checking anyway. 

so i would suggest the following additional warning:
```php
if (is_array($adapter)) {
    @trigger_error('passing options as the first constructor argument is deprecated in solarium 5.2 and will be removed in 6.x. use the third argument instead', E_USER_DEPRECATED);
    // ...
```

Reply:
@wickedOne I know changed it also for the event dispatcher.

Please check it again. I also added tests for the triggered deprecations using the Symfony phpunit bridge.

Reply:
@dmaicher yes, looks good to me!

still think an additional user deprecation when passing the options argument as the first argument is a good think to do as then it provides all the necessary warnings for moving from 5.2 to 6.0:
```php
        if (is_array($adapter)) {
            @trigger_error('passing options as the first constructor argument is deprecated in solarium 5.2 and will be removed in 6.0. use the third argument instead', E_USER_DEPRECATED);

            $options = $adapter;
        }
```
what do you think?

ping @mkalkbrenner

Reply:
But then they will get two deprecations. I think this one is clear enough?

https://github.com/solariumphp/solarium/pull/754/files#diff-f09ef20cbac899d71549b17bb68ceb2cR300

Reply:
yes, you're right